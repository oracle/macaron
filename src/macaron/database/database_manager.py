# Copyright (c) 2022 - 2024, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This DatabaseManager module handles the sqlite database connection."""
import collections.abc
import functools
import logging
import os
import typing

import sqlalchemy.exc
from sqlalchemy import create_engine, select
from sqlalchemy.orm import DeclarativeBase, Session

from macaron.config.defaults import defaults
from macaron.config.global_config import global_config
from macaron.database.views import create_view

logger: logging.Logger = logging.getLogger(__name__)


class ORMBase(DeclarativeBase):
    """ORM base class.

    To prevent Sphinx from rendering the docstrings in `DeclarativeBase`, make
    this docstring private:

    :meta private:
    """


class DatabaseManager:
    """This class handles and manages the connection to sqlite database during the session."""

    def __init__(self, db_path: str, base: type[DeclarativeBase] = ORMBase):
        """Initialize instance.

        Parameters
        ----------
        db_path : str
            The path to the target database.
        """
        self.engine = create_engine(f"sqlite+pysqlite:///{db_path}", echo=False)
        self.db_name = db_path
        self._base = base

    def create_tables(self) -> None:
        """
        Automatically create views for all tables known to _base.metadata.

        Creates all explicitly declared tables, and creates views proxying all tables beginning with an underscore.

        Note: this is specifically to allow the tables to be loaded into souffle:
            https://souffle-lang.github.io/directives#input-directive
        """
        try:
            self._base.metadata.create_all(self.engine, checkfirst=True)
            for table_name, table in self._base.metadata.tables.items():
                if table_name[0] == "_":
                    create_view(table_name[1:], self._base.metadata, select(table))
            self._base.metadata.create_all(self.engine, checkfirst=True)
        except sqlalchemy.exc.SQLAlchemyError as error:
            logger.error("Database error on create tables %s", error)


_T = typing.TypeVar("_T")
_P = typing.ParamSpec("_P")


class cache_return(typing.Generic[_T, _P]):  # pylint: disable=invalid-name # noqa: N801
    """The decorator to create a singleton DB session."""

    def __init__(self, function: collections.abc.Callable[_P, _T]) -> None:
        functools.update_wrapper(self, function)
        self.function = function

    def __call__(self, *args: _P.args, **kwargs: _P.kwargs) -> _T:
        """Store or get the cached function return value."""
        try:
            return self.return_value
        except AttributeError:
            self.return_value: _T = self.function(*args, **kwargs)  # pylint: disable=attribute-defined-outside-init
            return self.return_value

    def clear(self) -> None:
        """Remove the cached return value."""
        try:
            delattr(self, "return_value")
        except AttributeError:
            logger.debug("No cached return value to remove.")


@cache_return
def get_db_manager() -> DatabaseManager:
    """
    Get the database manager singleton object.

    Returns
    -------
    DatabaseManager
        The database manager singleton object.
    """
    db_path = os.path.join(global_config.output_path, defaults.get("database", "db_name", fallback="macaron.db"))
    db_man = DatabaseManager(db_path)
    return db_man


@cache_return
def get_db_session(session: Session | None = None) -> Session:
    """Get the current database session as a singleton object.

    This function expects to receive the Session object on the first
    call to cache it. The subsequent calls do not need to pass the
    Session object unless `get_db_session.clear()` is called.

    Parameters
    ----------
    Session | None
        The session object to be cached.

    Returns
    -------
    Session
        The current database session as a singleton object.

    Raises
    ------
    RuntimeError
        Happens if a database session is not created.
    """
    if session is None:
        raise RuntimeError("Failed to create a database session.")
    return session
