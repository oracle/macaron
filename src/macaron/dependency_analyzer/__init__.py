# Copyright (c) 2022 - 2023, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This package contains the dependency resolvers."""
