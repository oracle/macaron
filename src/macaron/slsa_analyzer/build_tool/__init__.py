# Copyright (c) 2022 - 2022, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""The build_tool package contains the supported build tools for Macaron."""

from .base_build_tool import BaseBuildTool
from .gradle import Gradle
from .maven import Maven

# The list of supported build tools. The order of the list determine the order
# in which each build tool is checked against the target repository.
BUILD_TOOLS: list[BaseBuildTool] = [Gradle(), Maven()]
