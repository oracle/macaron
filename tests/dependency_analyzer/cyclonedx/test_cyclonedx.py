# Copyright (c) 2023 - 2025, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module tests the CycloneDX helper functions."""
from pathlib import Path

import pytest
from cyclonedx.model.component import Component as CDXComponent

from macaron.database.table_definitions import Analysis, Component, RepoFinderMetadata, Repository
from macaron.dependency_analyzer.cyclonedx import (
    CycloneDXParserError,
    DependencyAnalyzer,
    DependencyInfo,
    deserialize_bom_json,
)
from macaron.slsa_analyzer.build_tool.base_build_tool import BaseBuildTool

RESOURCES_DIR = Path(__file__).parent.joinpath("resources")


def test_deserialize_bom_json(snapshot: list[str]) -> None:
    """Test deserializing a bom.json file."""
    # Deserialize a valid bom.json.
    # Note that serializing the Bom object and compare against the original bom would be an ideal test,
    # but cyclonedx took a really long time to finish serializing (could be a bug).
    # So, we compare the resolved components as a proxy.
    path = Path(RESOURCES_DIR, "valid_bom.json")
    bom = deserialize_bom_json(path)
    assert snapshot == [str(cmp.bom_ref) for cmp in bom.components if isinstance(cmp, CDXComponent)]

    # Deserialize a bom.json that does not exist.
    with pytest.raises(CycloneDXParserError):
        deserialize_bom_json(Path(RESOURCES_DIR, "does_not_exist"))

    # Deserialize an invalid bom.json.
    with pytest.raises(CycloneDXParserError):
        deserialize_bom_json(Path(RESOURCES_DIR, "invalid_bom.json"))

    # Deserialize an invalid JSON file.
    with pytest.raises(CycloneDXParserError):
        deserialize_bom_json(Path(RESOURCES_DIR, "invalid_json.json"))


@pytest.mark.parametrize(
    ("build_tool_name", "recursive"),
    [
        ("pip", False),
        ("pip", True),
        ("poetry", False),
        ("poetry", True),
    ],
)
def test_get_dep_components_python(
    snapshot: list, build_tools: dict[str, BaseBuildTool], build_tool_name: str, recursive: bool
) -> None:
    """Test retrieving dependencies as components."""
    # Path to the root bom.json.
    root_bom_path = Path(RESOURCES_DIR, "bom_requests.json")

    dep_analyzer = build_tools[build_tool_name].get_dep_analyzer()
    component = Component(
        purl="pkg:pypi/requests@2.31.0",
        analysis=Analysis(),
        repository=Repository(complete_name="github.com/psf/requests", fs_path=""),
        repo_finder_metadata=RepoFinderMetadata(),
    )

    # Path to the sub-project bom.json files.
    result_bom_refs = sorted(
        res.bom_ref.value
        for res in dep_analyzer.get_dep_components(
            target_component=component, root_bom_path=root_bom_path, recursive=recursive
        )
        if res.bom_ref.value
    )
    assert snapshot == result_bom_refs


@pytest.mark.parametrize(
    "build_tool_name",
    [
        "pip",
        "poetry",
    ],
)
def test_convert_components_to_artifacts_python(
    snapshot: dict[str, DependencyInfo], build_tools: dict[str, BaseBuildTool], build_tool_name: str
) -> None:
    """Test converting CycloneDX components using internal artifact representation."""
    # Path to the root bom.json.
    root_bom_path = Path(RESOURCES_DIR, "bom_requests.json")

    dep_analyzer = build_tools[build_tool_name].get_dep_analyzer()
    component = Component(
        purl="pkg:pypi/requests@2.31.0",
        analysis=Analysis(),
        repository=Repository(complete_name="github.com/psf/requests", fs_path=""),
        repo_finder_metadata=RepoFinderMetadata(),
    )

    # Pass the root bom.json.
    result = dep_analyzer.convert_components_to_artifacts(
        dep_analyzer.get_dep_components(target_component=component, root_bom_path=root_bom_path, recursive=True),
        purl_type=component.namespace,
    )
    assert snapshot == result


@pytest.mark.parametrize(
    "name",
    [
        "bom_no_group.json",
        "bom_no_version.json",
    ],
)
def test_low_quality_bom(snapshot: dict[str, DependencyInfo], name: str) -> None:
    """Test converting CycloneDX components when the quality of the BOM file is poor.

    E.g., a component misses a version or group.
    """
    # Path to the BOM file.
    bom_path = Path(RESOURCES_DIR, name)

    component = Component(
        purl="pkg:maven/com.amazonaws/aws-lambda-java-events@3.11.0?type=jar",
        analysis=Analysis(),
        repository=Repository(complete_name="github.com/aws/aws-lambda-java-libs", fs_path=""),
        repo_finder_metadata=RepoFinderMetadata(),
    )
    result = DependencyAnalyzer.get_deps_from_sbom(bom_path, target_component=component)
    assert snapshot == result


def test_multiple_versions(snapshot: dict[str, DependencyInfo]) -> None:
    """Test converting CycloneDX components when there are multiple artifacts.

    Based on semantic versioning, version strings can contain alphabet characters.
    """
    # Path to the BOM file.
    bom_path = Path(RESOURCES_DIR, "bom_multi_versions.json")
    component = Component(
        purl="pkg:maven/com.amazonaws/aws-lambda-java-events@3.11.0?type=jar",
        analysis=Analysis(),
        repository=Repository(complete_name="github.com/aws/aws-lambda-java-libs", fs_path=""),
        repo_finder_metadata=RepoFinderMetadata(),
    )
    result = DependencyAnalyzer.get_deps_from_sbom(bom_path, target_component=component)
    assert snapshot == result


@pytest.mark.parametrize(
    ("purl_type", "group", "name", "version", "expected_purl"),
    [
        ("pypi", None, "django", "5.0.6", "pkg:pypi/django@5.0.6"),
        ("maven", "org.apache", "maven", "1.0.0", "pkg:maven/org.apache/maven@1.0.0"),
        ("maven", "com.google", "guava", "32.1.2-jre", "pkg:maven/com.google/guava@32.1.2-jre"),
    ],
)
def test_get_purl_from_cdx_component(
    purl_type: str,
    group: str | None,
    name: str,
    version: str,
    expected_purl: str,
) -> None:
    """Test constructing a PackageURL from a CycloneDX component."""
    component = CDXComponent(group=group, name=name, version=version)
    assert (
        str(DependencyAnalyzer.get_purl_from_cdx_component(component=component, purl_type=purl_type)) == expected_purl
    )
