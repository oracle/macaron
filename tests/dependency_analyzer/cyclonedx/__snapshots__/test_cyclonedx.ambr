# serializer version: 1
# name: test_convert_components_to_artifacts_python[pip]
  dict({
    'None:certifi': dict({
      'available': <SCMStatus.AVAILABLE: 'AVAILABLE'>,
      'note': '',
      'purl': PackageURL(
        name='certifi',
        namespace=None,
        qualifiers=dict({
        }),
        subpath=None,
        type='pypi',
        version='2024.2.2',
      ),
      'url': 'https://github.com/certifi/python-certifi',
    }),
    'None:charset-normalizer': dict({
      'available': <SCMStatus.AVAILABLE: 'AVAILABLE'>,
      'note': '',
      'purl': PackageURL(
        name='charset-normalizer',
        namespace=None,
        qualifiers=dict({
        }),
        subpath=None,
        type='pypi',
        version='3.3.2',
      ),
      'url': 'https://github.com/Ousret/charset_normalizer',
    }),
    'None:idna': dict({
      'available': <SCMStatus.AVAILABLE: 'AVAILABLE'>,
      'note': '',
      'purl': PackageURL(
        name='idna',
        namespace=None,
        qualifiers=dict({
        }),
        subpath=None,
        type='pypi',
        version='3.7',
      ),
      'url': 'https://github.com/kjd/idna',
    }),
    'None:pip': dict({
      'available': <SCMStatus.AVAILABLE: 'AVAILABLE'>,
      'note': '',
      'purl': PackageURL(
        name='pip',
        namespace=None,
        qualifiers=dict({
        }),
        subpath=None,
        type='pypi',
        version='23.2.1',
      ),
      'url': 'https://github.com/pypa/pip',
    }),
    'None:urllib3': dict({
      'available': <SCMStatus.AVAILABLE: 'AVAILABLE'>,
      'note': '',
      'purl': PackageURL(
        name='urllib3',
        namespace=None,
        qualifiers=dict({
        }),
        subpath=None,
        type='pypi',
        version='2.2.1',
      ),
      'url': 'https://github.com/urllib3/urllib3',
    }),
  })
# ---
# name: test_convert_components_to_artifacts_python[poetry]
  dict({
    'None:certifi': dict({
      'available': <SCMStatus.AVAILABLE: 'AVAILABLE'>,
      'note': '',
      'purl': PackageURL(
        name='certifi',
        namespace=None,
        qualifiers=dict({
        }),
        subpath=None,
        type='pypi',
        version='2024.2.2',
      ),
      'url': 'https://github.com/certifi/python-certifi',
    }),
    'None:charset-normalizer': dict({
      'available': <SCMStatus.AVAILABLE: 'AVAILABLE'>,
      'note': '',
      'purl': PackageURL(
        name='charset-normalizer',
        namespace=None,
        qualifiers=dict({
        }),
        subpath=None,
        type='pypi',
        version='3.3.2',
      ),
      'url': 'https://github.com/Ousret/charset_normalizer',
    }),
    'None:idna': dict({
      'available': <SCMStatus.AVAILABLE: 'AVAILABLE'>,
      'note': '',
      'purl': PackageURL(
        name='idna',
        namespace=None,
        qualifiers=dict({
        }),
        subpath=None,
        type='pypi',
        version='3.7',
      ),
      'url': 'https://github.com/kjd/idna',
    }),
    'None:pip': dict({
      'available': <SCMStatus.AVAILABLE: 'AVAILABLE'>,
      'note': '',
      'purl': PackageURL(
        name='pip',
        namespace=None,
        qualifiers=dict({
        }),
        subpath=None,
        type='pypi',
        version='23.2.1',
      ),
      'url': 'https://github.com/pypa/pip',
    }),
    'None:urllib3': dict({
      'available': <SCMStatus.AVAILABLE: 'AVAILABLE'>,
      'note': '',
      'purl': PackageURL(
        name='urllib3',
        namespace=None,
        qualifiers=dict({
        }),
        subpath=None,
        type='pypi',
        version='2.2.1',
      ),
      'url': 'https://github.com/urllib3/urllib3',
    }),
  })
# ---
# name: test_deserialize_bom_json
  list([
    'pkg:maven/ch.qos.logback/logback-classic@1.2.11?type=jar',
    'pkg:maven/ch.qos.logback/logback-core@1.2.11?type=jar',
    'pkg:maven/ch.qos.logback.contrib/logback-json-classic@0.1.5?type=jar',
    'pkg:maven/ch.qos.logback.contrib/logback-json-core@0.1.5?type=jar',
    'pkg:maven/com.amazon.alexa/ask-sdk-core@2.49.0?type=jar',
    'pkg:maven/com.amazon.alexa/ask-sdk-lambda-support@2.49.0?type=jar',
    'pkg:maven/com.amazon.alexa/ask-sdk-model@1.43.0?type=jar',
    'pkg:maven/com.amazon.alexa/ask-sdk-model-runtime@1.0.5?type=jar',
    'pkg:maven/com.amazon.alexa/ask-sdk-runtime@2.49.0?type=jar',
    'pkg:maven/com.amazonaws/aws-java-sdk-core@1.12.382?type=jar',
    'pkg:maven/com.amazonaws/aws-java-sdk-lambda@1.12.382?type=jar',
    'pkg:maven/com.amazonaws/aws-lambda-java-core@1.2.2?type=jar',
    'pkg:maven/com.amazonaws/aws-lambda-java-events@3.11.0?type=jar',
    'pkg:maven/com.amazonaws/jmespath-java@1.12.382?type=jar',
    'pkg:maven/com.amazonaws.serverless/aws-serverless-java-container-core@1.9.1?type=jar',
    'pkg:maven/com.fasterxml.jackson.core/jackson-annotations@2.14.1?type=jar',
    'pkg:maven/com.fasterxml.jackson.core/jackson-core@2.14.1?type=jar',
    'pkg:maven/com.fasterxml.jackson.core/jackson-databind@2.14.1?type=jar',
    'pkg:maven/com.fasterxml.jackson.dataformat/jackson-dataformat-cbor@2.14.1?type=jar',
    'pkg:maven/com.fasterxml.jackson.datatype/jackson-datatype-jdk8@2.14.1?type=jar',
    'pkg:maven/com.fasterxml.jackson.datatype/jackson-datatype-jsr310@2.14.1?type=jar',
    'pkg:maven/com.fizzed/rocker-runtime@1.3.0?type=jar',
    'pkg:maven/com.googlecode.javaewah/JavaEWAH@1.1.7?type=jar',
    'pkg:maven/com.typesafe/config@1.4.1?type=jar',
    'pkg:maven/commons-codec/commons-codec@1.11?type=jar',
    'pkg:maven/commons-codec/commons-codec@1.15?type=jar',
    'pkg:maven/commons-fileupload/commons-fileupload@1.4?type=jar',
    'pkg:maven/commons-io/commons-io@2.11.0?type=jar',
    'pkg:maven/commons-io/commons-io@2.7?type=jar',
    'pkg:maven/commons-logging/commons-logging@1.2?type=jar',
    'pkg:maven/io.github.java-diff-utils/java-diff-utils@4.10?type=jar',
    'pkg:maven/io.micronaut/micronaut-aop@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-buffer-netty@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-context@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-core@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-core-reactive@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-function@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-function-client@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-http@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-http-client@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-http-client-core@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-http-netty@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-http-server@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-inject@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-jackson-core@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-jackson-databind@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-json-core@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-router@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-runtime@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-validation@3.8.0?type=jar',
    'pkg:maven/io.micronaut/micronaut-websocket@3.8.0?type=jar',
    'pkg:maven/io.micronaut.aws/aws-alexa@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/aws-alexa-httpserver@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/aws-apigateway@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/aws-bom@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/aws-cdk@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/aws-cloudwatch-logging@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/aws-common@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/aws-distributed-configuration@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/aws-parameter-store@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/aws-sdk-v1@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/aws-sdk-v2@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/aws-secretsmanager@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/aws-service-discovery@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/aws-ua@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/function-aws@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/function-aws-alexa@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/function-aws-api-proxy@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/function-aws-api-proxy-test@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/function-aws-custom-runtime@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/function-aws-test@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/function-client-aws@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/test-suite@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/test-suite-aws-sdk-v2@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/test-suite-groovy@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/test-suite-http-server-tck-function-aws-api-proxy@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.aws/test-suite-kotlin@4.0.0-SNAPSHOT?type=jar',
    'pkg:maven/io.micronaut.discovery/micronaut-discovery-client@3.2.0?type=jar',
    'pkg:maven/io.micronaut.serde/micronaut-serde-api@1.5.0?type=jar',
    'pkg:maven/io.micronaut.serde/micronaut-serde-jackson@1.5.0?type=jar',
    'pkg:maven/io.micronaut.serde/micronaut-serde-support@1.5.0?type=jar',
    'pkg:maven/io.micronaut.starter/micronaut-starter-api@3.8.0?type=jar',
    'pkg:maven/io.micronaut.starter/micronaut-starter-core@3.8.0?type=jar',
    'pkg:maven/io.micronaut.test/micronaut-test-core@3.8.0?type=jar',
    'pkg:maven/io.micronaut.test/micronaut-test-junit5@3.8.0?type=jar',
    'pkg:maven/io.micronaut.testresources/micronaut-test-resources-build-tools@1.2.3?type=jar',
    'pkg:maven/io.netty/netty-buffer@4.1.86.Final?type=jar',
    'pkg:maven/io.netty/netty-codec@4.1.86.Final?type=jar',
    'pkg:maven/io.netty/netty-codec-http@4.1.86.Final?type=jar',
    'pkg:maven/io.netty/netty-codec-http2@4.1.86.Final?type=jar',
    'pkg:maven/io.netty/netty-codec-socks@4.1.86.Final?type=jar',
    'pkg:maven/io.netty/netty-common@4.1.86.Final?type=jar',
    'pkg:maven/io.netty/netty-handler@4.1.86.Final?type=jar',
    'pkg:maven/io.netty/netty-handler-proxy@4.1.86.Final?type=jar',
    'pkg:maven/io.netty/netty-resolver@4.1.86.Final?type=jar',
    'pkg:maven/io.netty/netty-transport@4.1.86.Final?type=jar',
    'pkg:maven/io.netty/netty-transport-classes-epoll@4.1.86.Final?type=jar',
    'pkg:maven/io.netty/netty-transport-native-unix-common@4.1.86.Final?type=jar',
    'pkg:maven/io.projectreactor/reactor-core@3.5.0?type=jar',
    'pkg:maven/io.swagger.core.v3/swagger-annotations@2.2.7?type=jar',
    'pkg:maven/jakarta.annotation/jakarta.annotation-api@2.1.1?type=jar',
    'pkg:maven/jakarta.inject/jakarta.inject-api@2.0.1?type=jar',
    'pkg:maven/javax.annotation/javax.annotation-api@1.3.2?type=jar',
    'pkg:maven/javax.inject/javax.inject@1?type=jar',
    'pkg:maven/javax.servlet/javax.servlet-api@4.0.1?type=jar',
    'pkg:maven/javax.validation/validation-api@2.0.1.Final?type=jar',
    'pkg:maven/javax.ws.rs/javax.ws.rs-api@2.1?type=jar',
    'pkg:maven/joda-time/joda-time@2.6?type=jar',
    'pkg:maven/joda-time/joda-time@2.8.1?type=jar',
    'pkg:maven/org.apache.commons/commons-compress@1.21?type=jar',
    'pkg:maven/org.apache.httpcomponents/httpclient@4.5.13?type=jar',
    'pkg:maven/org.apache.httpcomponents/httpcore@4.4.13?type=jar',
    'pkg:maven/org.apache.httpcomponents/httpmime@4.5.13?type=jar',
    'pkg:maven/org.apache.logging.log4j/log4j-api@2.19.0?type=jar',
    'pkg:maven/org.apache.logging.log4j/log4j-core@2.19.0?type=jar',
    'pkg:maven/org.apache.logging.log4j/log4j-slf4j-impl@2.19.0?type=jar',
    'pkg:maven/org.eclipse.jetty/jetty-http@9.4.50.v20221201?type=jar',
    'pkg:maven/org.eclipse.jetty/jetty-io@9.4.50.v20221201?type=jar',
    'pkg:maven/org.eclipse.jetty/jetty-server@9.4.50.v20221201?type=jar',
    'pkg:maven/org.eclipse.jetty/jetty-util@9.4.50.v20221201?type=jar',
    'pkg:maven/org.eclipse.jgit/org.eclipse.jgit@5.11.1.202105131744-r?type=jar',
    'pkg:maven/org.jetbrains/annotations@13.0?type=jar',
    'pkg:maven/org.jetbrains/annotations@19.0.0?type=jar',
    'pkg:maven/org.jetbrains.kotlin/kotlin-stdlib@1.7.21?type=jar',
    'pkg:maven/org.jetbrains.kotlin/kotlin-stdlib-common@1.7.21?type=jar',
    'pkg:maven/org.jetbrains.kotlin/kotlin-stdlib-jdk7@1.7.21?type=jar',
    'pkg:maven/org.jetbrains.kotlin/kotlin-stdlib-jdk8@1.7.21?type=jar',
    'pkg:maven/org.junit.jupiter/junit-jupiter-api@5.9.1?type=jar',
    'pkg:maven/org.junit.platform/junit-platform-commons@1.9.1?type=jar',
    'pkg:maven/org.opentest4j/opentest4j@1.2.0?type=jar',
    'pkg:maven/org.reactivestreams/reactive-streams@1.0.4?type=jar',
    'pkg:maven/org.slf4j/jcl-over-slf4j@1.7.36?type=jar',
    'pkg:maven/org.slf4j/slf4j-api@1.7.36?type=jar',
    'pkg:maven/org.yaml/snakeyaml@1.33?type=jar',
    'pkg:maven/software.amazon.awscdk/aws-cdk-lib@2.59.0?type=jar',
    'pkg:maven/software.amazon.awscdk/cdk-asset-awscli-v1@2.2.52?type=jar',
    'pkg:maven/software.amazon.awscdk/cdk-asset-kubectl-v20@2.1.1?type=jar',
    'pkg:maven/software.amazon.awscdk/cdk-asset-node-proxy-agent-v5@2.0.42?type=jar',
    'pkg:maven/software.amazon.awssdk/annotations@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/apache-client@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/auth@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/aws-core@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/aws-json-protocol@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/cloudwatchlogs@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/endpoints-spi@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/http-client-spi@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/json-utils@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/metrics-spi@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/netty-nio-client@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/profiles@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/protocol-core@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/regions@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/sdk-core@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/secretsmanager@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/servicediscovery@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/ssm@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/third-party-jackson-core@2.19.14?type=jar',
    'pkg:maven/software.amazon.awssdk/utils@2.19.14?type=jar',
    'pkg:maven/software.amazon.eventstream/eventstream@1.0.1?type=jar',
    'pkg:maven/software.amazon.ion/ion-java@1.0.2?type=jar',
    'pkg:maven/software.amazon.jsii/jsii-runtime@1.73.0?type=jar',
    'pkg:maven/software.constructs/constructs@10.1.232?type=jar',
  ])
# ---
# name: test_get_dep_components_python[pip-False]
  list([
    'certifi==2024.2.2',
    'charset-normalizer==3.3.2',
    'idna==3.7',
    'urllib3==2.2.1',
  ])
# ---
# name: test_get_dep_components_python[pip-True]
  list([
    'certifi==2024.2.2',
    'charset-normalizer==3.3.2',
    'idna==3.7',
    'pip==23.2.1',
    'urllib3==2.2.1',
  ])
# ---
# name: test_get_dep_components_python[poetry-False]
  list([
    'certifi==2024.2.2',
    'charset-normalizer==3.3.2',
    'idna==3.7',
    'urllib3==2.2.1',
  ])
# ---
# name: test_get_dep_components_python[poetry-True]
  list([
    'certifi==2024.2.2',
    'charset-normalizer==3.3.2',
    'idna==3.7',
    'pip==23.2.1',
    'urllib3==2.2.1',
  ])
# ---
# name: test_low_quality_bom[bom_no_group.json]
  dict({
    'None:joda-time': dict({
      'available': <SCMStatus.AVAILABLE: 'AVAILABLE'>,
      'note': '',
      'purl': PackageURL(
        name='joda-time',
        namespace='joda-time',
        qualifiers=dict({
          'type': 'jar',
        }),
        subpath=None,
        type='maven',
        version='2.6',
      ),
      'url': 'https://github.com/JodaOrg/joda-time',
    }),
  })
# ---
# name: test_low_quality_bom[bom_no_version.json]
  dict({
    'joda-time:joda-time': dict({
      'available': <SCMStatus.AVAILABLE: 'AVAILABLE'>,
      'note': '',
      'purl': PackageURL(
        name='joda-time',
        namespace='joda-time',
        qualifiers=dict({
          'type': 'jar',
        }),
        subpath=None,
        type='maven',
        version='2.6',
      ),
      'url': 'https://github.com/JodaOrg/joda-time',
    }),
  })
# ---
# name: test_multiple_versions
  dict({
    'joda-time:joda-time': dict({
      'available': <SCMStatus.AVAILABLE: 'AVAILABLE'>,
      'note': '',
      'purl': PackageURL(
        name='joda-time',
        namespace='joda-time',
        qualifiers=dict({
          'type': 'jar',
        }),
        subpath=None,
        type='maven',
        version='2.6',
      ),
      'url': 'https://github.com/JodaOrg/joda-time',
    }),
  })
# ---
